smolyay/smolyak.py

```diff
@@ -0,0 +1,141 @@
+import numpy as np
+import scipy.special
+
+
+class IndexGrid:
+    """Generate index of grid points.
+
+    Attributes
+    ----------
+    dimension : int
+        Number of independent variables.
+    exactness : int
+        Level of exactness of the approximation.
+    index_per_level : list
+        Number of indexes at each level.
+    """
+
+    def __init__(self, dimension, exactness, index_per_level):
+        self.dimension = dimension
+        self.exactness = exactness
+        self.index_per_level = index_per_level
+
+        if len(index_per_level) < exactness + 1:
+            raise IndexError(
+                "index_per_level must be an array with a"
+                "length of at least {}"
+                .format(exactness + 1))
+
+    def level_indexes(self):
+        """Generate indexes of levels.
+
+        Returns
+        -------
+        level_indexes : list
+            Unique indexes at each level.
+
+        """
+        # creating an array to store unique indexes in each level
```

Comment:
I think this whole function can all be done in a more readable way like this
```py
# cumulative sum to get end index of each level, up to exactness+1
end_levels = numpy.cumsum(self.index_per_level[:self.exactness+1])
# create ranges of indexes at each level
level_indexes = [list(range(end-n, end)) for end, n in zip(end_levels, self.index_per_level)]
return level_indexes
```

Reply:
Done.